Open Judge System/OJS.Services/Business/Problems/ProblemsBusinessService.cs

```diff
@@ -38,5 +63,52 @@ public void RetestById(int id)
                 scope.Complete();
             }
         }
+
+        public void DeleteById(int id)
+        {
+            var problem = this.problemsData
+                .GetByIdQuery(id)
+                .Select(p => new
+                {
+                    p.ProblemGroupId,
+                    p.ProblemGroup.ContestId
+                })
+                .FirstOrDefault();
+
+            if (problem == null)
+            {
+                return;
+            }
+
+            this.testRunsData.DeleteByProblem(id);
+
+            this.problemResourcesData.DeleteByProblem(id);
+
+            this.submissionsData.DeleteByProblem(id);
+
+            this.problems.Delete(id);
+            this.problems.SaveChanges();
+
+            if (!this.contestsData.IsOnlineById(problem.ContestId))
+            {
+                this.problemGroupsBusiness.DeleteById(problem.ProblemGroupId.Value);
+            }
+        }
+
+        public void DeleteByContest(int contestId)
+        {
+            this.testRunsData.DeleteByContest(contestId);
+
+            this.problemResourcesData.DeleteByContest(contestId);
+
+            this.submissionsData.DeleteByContest(contestId);
+
+            this.problems.Delete(p => p.ProblemGroup.ContestId == contestId);
+
```

Comment:
On the other call to problems repo you call SaveChanges, but you don't call it here?